### PR TITLE
[Elasticsearch] contextual support

### LINKED ISSUE #249: Contextual / runtime params support
**Description**  

In my use cases, there exist some use-cases where I want to have the ability to enrich the search with params which are not under user's direct control.

I'll write in Elasticsearch form because that allows me to be more precise, but the feature should probably be supported in core. Also, imagine this is a RESTful API.

# Contextual params in mappings (simple)

If I want to support multi-language, I can have multiple Elasticsearch indices `articles_en`, `articles_hr`, `articles_nl`.

This could be something like

```
Accept-Language: en
GET /api/articles?search=title:apple

Accept-Language: hr
GET /api/articles?search=title:jabuka
```

What I'd do is map it to an Elasticsearch index like this

```
title => /articles_{locale}#title
```

and then (somehow?) inject the `locale` param which is known at runtime only.

It would yield

```
POST /articles_en/_search
{
    query: {
        match: {
            title: apple
        }
    }
}
```

# Contextual params in mappings (advanced)

Variation of the case above: let's say I want each user to be able to only search their own articles. They would send their access token and an identical query, but get different results.

Example usage:

```
Authorization: Bearer user1
GET /api/articles?search=title:apple

Authorization: Bearer user2
GET /api/articles?search=title:apple
```

would yield

```
POST /articles_en/_search
{
    query: {
        match: {
            title: apple,
            user: user1 # (or user2)
        }
    }
}
```

The additional injected condition could either be static (example: `published: true`) or dynamic, like in this case.

This looks a lot like the primary condition thing we already added, but it should only be applied if this specific field is applied too.

# Contextual fields (very advanced)

So, a variation of the previous example: let's say `user1` and `user2` have **different fields** to search against (for example, EVA model).

By creating a `has_child` query (implemented in #248), one user could have access to a specific child structure (since he's the owner) while the other does not at all, so the additional fields are added at runtime.

Example:

```
Authorization: Bearer user1
GET /api/articles?search=note:apple
HTTP 200

Authorization: Bearer user2
GET /api/articles?search=note:apple
HTTP 400 (note param not available at all to this user)
```

Combining all these example, a mapping would look something like this:

```
note => /articles_{locale}#extended>values.values_{user}.note
```

and would yield

```
POST /articles_en/_search
{
    query: {
        has_child: {
            type: extended,
            query: {
               match: {
                   values.values_user1.note: apple
                   user: user1 # (or user2)
            }
        }
    }
}
```

**Example**  

Given in the description.

### PR DESCRIPTION
**Pull request by dkarlovi**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #249 
| License       | MIT
| Doc PR        | N/A

Closes #249. Solves cases 1) and 2) from there. Case 3) is handled by decorating the factory, all use cases work well in my app.

### CODE REVIEW

**Comment by dkarlovi** [comment]
@sstok  this is already solving points 1) and 2) from #249, do you have any objections about the direction?

**Comment by dkarlovi** [comment]
BTW this is a snippet of my APIP mapping:

```xml
                <attribute name="_any">
                    <attribute name="elasticsearch">
                        <attribute name="mappings">

                           <!-- case 1) and 2) from #249 -->
                            <attribute name="origin">
                                <attribute name="property">/app_{locale}_article/article#origin.identifier</attribute>
                                <attribute name="conditions">
                                    <attribute>
                                        <attribute name="property">/app_{locale}_article/article#origin.client</attribute>
                                        <attribute name="value">{client_id}</attribute>
                                    </attribute>
                                </attribute>
                            </attribute>

                           <!-- to be case 3) from #249 (needs to be added at runtime) -->
                            <attribute name="att_note">
                                <attribute name="property">/app_{locale}_article/article#extended_attribute_bag_client_user>extended_attribute_bag_client_user.values_{client_id}.note</attribute>
                                <attribute name="conditions">
                                    <attribute>
                                        <attribute name="property">/app_{locale}_article/article#extended_attribute_bag_client_user>user</attribute>
                                        <attribute name="value">{user_id}</attribute>
                                    </attribute>
                                </attribute>
                            </attribute>
                        </attribute>
                </attribute>
            </attribute>

**Comment by dkarlovi** [comment]
Remaining:
1. "unpack" the conditions supplied with the mapping and apply them too if non-empty
2. figure out how to create a mapping dynamically (this will probably be tricky)

**Comment by dkarlovi** [comment]
@sstok this works for contextual conditions already, you can see both runtime properties and contextual conditions being assigned.

The only thing that comes to mind might be: currently it only allows for simple comparison with AND in contextual conditions, which might be too restrictive, but it also might be fine (so it behaves like `AND additional_field = :value` in SQL).

IMO we could clean up this PR and merge here, adding runtime mappings could be a separate PR.

WDYT?

**Comment by dkarlovi** [comment]
This works as-is, ready for review.

**Comment by dkarlovi** [comment]
Yay, all green! \o/

@sstok do you think you'd have time to review?

**Comment by dkarlovi** [comment]
> are you going to continue in this pull request or in a follow-up?

Not really much to follow-up except #252, but I think that's fine as a separate PR, since it would require some discussion.

>  some tests would be nice

Adding now. :+1:

File: lib/Symfony/SearchBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -54,5 +54,8 @@
         <service id="rollerworks_search.cache.adapter.array" class="Symfony\Component\Cache\Adapter\ArrayAdapter">
             <argument>0</argument> <!-- default lifetime -->
         </service>
+
+        <!-- ParameterBag -->
+        <service alias="rollerworks_search.parameter_bag" id="Rollerworks\Component\Search\ParameterBag"/>
```
Comment (dkarlovi) [inline lib/Symfony/SearchBundle/Resources/config/services.xml]: It seems you cannot alias the service like this?

Comment (sstok) [reply lib/Symfony/SearchBundle/Resources/config/services.xml]: Only if the class is actually registered as a service.

**Comment by dkarlovi** [comment]
@sstok it should be ready now, tests added from the other branch and it all works here and in my app.

**Comment by sstok** [comment]
Can I merge? 👍

**Comment by dkarlovi** [comment]
Yup! \o/

**Comment by sstok** [comment]
Thank you @dkarlovi

**Comment by dkarlovi** [comment]
Thank you for the merge. :+1:

**Pull request merged**